### PR TITLE
Enable use of spot instances in AKS

### PR DESCRIPTION
**Pull request by jimgus**
This PR makes it possible make use of Spot instance node pools in AKS. Spot instance pools can be used in conjunction with "standard" in order to assure capacity. That is the reason that this PR makes it possible to configure a "priority" expander in addition to the "least-waste" expander previously used. The "priority" expander allows you to put priority on the different node pools. If a "standard" pool and a "spot" pool are configured, it is possible to put a higher priority on the "spot" pool and hence this pool will be expanded by the autoscaler it there are spot instances available. Otherwise will the "standard" pool be expanded. Setups like this can be used in dev/qa environments in order to be able to reduce costs.

Spot instances will always be tainted with `kubernetes.azure.com/scalesetpriority=spot:NoSchedule`. This taint is removed by a OPA Gatekeeper mutating webhook in order to be able to schedule workloads on spot nodes.

### CODE REVIEW

File: modules/kubernetes/aks-core/modules.tf [hunk 1]
```diff
@@ -105,6 +107,7 @@ module "fluxcd_v2_azure_devops" {
 }
 
 module "fluxcd_v2_github" {
+
```
Comment (NissesSenap) [inline modules/kubernetes/aks-core/modules.tf:110]: why \n?

**Review by NissesSenap — COMMENTED** [review]
Minor comment and a general discussion point.
Other then that it LGTM, looking forward to using this feature.

**Comment by phillebaba** [comment]
> Spot instances will always be tainted with kubernetes.azure.com/scalesetpriority=spot:NoSchedule. This taint is removed by a OPA Gatekeeper mutating webhook in order to be able to schedule workloads on spot nodes.

Are we removing the taint on the node or adding the toleration to all pods?

**Comment by NissesSenap** [comment]
@phillebaba 
We are removing the taint from the nodes. It felt like it was the easiest solution.

**Comment by phillebaba** [comment]
My only thought is that we should keep the taint on the Nodes and instead just mutate all Pods to tolerate the taint. This way it is technically possible to opt out of spot instances. Thoughts on this?

**Comment by jimgus** [comment]
> 

It's a tricky thing to get that toleration on all pods, especially considering we have the `excluded_namespaces` for opa-gatekeeper. Not sure it is a valid use case either. I suggest that we do that when we see an actual need for it.

File: modules/azure/aks/aks.tf [hunk 2]
```diff
@@ -82,7 +89,29 @@ resource "azurerm_kubernetes_cluster" "this" {
   }
 }
 
+provider "kubectl" {
+  host                   = azurerm_kubernetes_cluster.this.kube_config[0].host
+  client_certificate     = base64decode(azurerm_kubernetes_cluster.this.kube_admin_config[0].client_certificate)
+  client_key             = base64decode(azurerm_kubernetes_cluster.this.kube_admin_config[0].client_key)
+  cluster_ca_certificate = base64decode(azurerm_kubernetes_cluster.this.kube_admin_config[0].cluster_ca_certificate)
+  load_config_file       = false
+}
```
[the remark this answers is not in the document]
Comment (jimgus) [reply modules/azure/aks/aks.tf:98]: Is that really possible in this case?

File: CHANGELOG.md [hunk 3]
```diff
@@ -21,6 +21,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 - [#897](https://github.com/XenitAB/terraform-modules/pull/897) Add Datadog APM ignore resources
 - [#921](https://github.com/XenitAB/terraform-modules/pull/921) Add Azure Alerts if no data gets sent to log storage account.
+- [#922](https://github.com/XenitAB/terraform-modules/pull/922) Enable use of spot instances in AKS
+
+### Changed
+
+- [#920](https://github.com/XenitAB/terraform-modules/pull/920) Increase default AKS audit retention to 365 days.
+>>>>>>> f91050c9 (Enable use of spot instances in AKS)
```
Comment (phillebaba) [inline CHANGELOG.md:29]: Seems to be an artifact of rebasing

File: modules/azure/aks/aks.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (phillebaba) [reply modules/azure/aks/aks.tf:98]: This has to end up in AKS Core. I tried doing something similar before but it broke module dependencies by defining providers in the module-

[the remark this answers is not in the document]
Comment (jimgus) [reply modules/azure/aks/aks.tf]: done

File: modules/azure/aks/aks.tf [hunk 4]
```diff
@@ -82,7 +89,29 @@ resource "azurerm_kubernetes_cluster" "this" {
...
+
+resource "kubectl_manifest" "priority_expander" {
```
[the remark this answers is not in the document]
Comment (jimgus) [reply modules/azure/aks/aks.tf]: done

File: modules/kubernetes/aks-core/modules.tf (hunk 1 above)
Comment (jimgus) [reply modules/kubernetes/aks-core/modules.tf]: removed

**Review by NissesSenap — APPROVED** [review]
LGTM

**Review by phillebaba — APPROVED** [review]
LGTM great work!

**Pull request merged**